### ISSUE TITLE
Clarify type requirements of operator->

### ISSUE THREAD

**Issue by offa** [offa/scope-guard#22]
The `operator->` requires a type that satisfies (*§ 8.4.1/11*):

 ```cpp
is_pointer<R>::value && (is_class<R>::value || is_union<R>::value)
```

This however doesn't work, `R` as either a Pointer type or a class / union.

**Comment by offa** [comment]
**Ref.** #23

**Comment by offa** [comment]
Corrected in P0052 (#26): 

```cpp
is_pointer<R>::value &&
(is_class<remove_pointer_t<R>>::value || is_union<remove_pointer_t<R>>::value)
```

This is also how it's implemented at the moment. Except, using `decay()`.
